Add ability to set dataset create-time allocation policy

jmert opened the pull request:
This PR adds the `H5D_alloc_time_t` constants and getter/setter functions to control the dataset allocation policy, as described by https://portal.hdfgroup.org/display/HDF5/H5P_SET_ALLOC_TIME.

This is built upon #644 since I found that bug while learning about early vs late allocation, and I've got a use case where I'd like to try using early allocation with mmaped arrays. I also see #632 is proposed to overhaul the keyword handling a bunch, so I can rebase this change on that if it lands first.

~~I haven't added any tests that the `get_alloc_time()` function works since I didn't see any obvious place them, but I could create a new file/testset just for setting and getting property values if desired (which will probably be needed if more of the properties such as fill value and fill time are hooked up later).~~ Added in a new commit.

musm commented:
Will need to be rebased otherwise LGTM

musm commented:
Should we add some simple docs  on this at https://github.com/JuliaIO/HDF5.jl/blob/master/docs/src/index.md

jmert commented:
Rebased to use properties, and I've added a short description to the Memory Mapping section on why you might want to change the allocation-time policy.

One thing I've left out is any mention of the fact that early allocation in some cases is still very cheap due to OS/filesystem support for [sparse files](https://en.wikipedia.org/wiki/Sparse_file) because (a) that seems kind of out of the scope of this package's documentation (even thought it's quite useful to know), and (b) I have no clue (since I only use Linux) how sparse files are handled across Mac/Windows filesystems in concert with whatever HDF5 library is used there.

musm commented:
Looks great, 


> One thing I've left out is any mention of the fact that early allocation in some cases is still very cheap due to OS/filesystem support for sparse files because (a) that seems kind of out of the scope of this package's documentation (even though it's quite useful to know), and (b) I have no clue (since I only use Linux) how sparse files are handled across Mac/Windows filesystems in concert with whatever HDF5 library is used there.

Right, it does seem a little out of scope. It's always possible to add a footnote in that section regarding this important point.
